src/transformers/generation/stopping_criteria.py

```diff
@@ -126,6 +129,167 @@ def __call__(self, input_ids: torch.LongTensor, scores: torch.FloatTensor, **kwa
+        # Flip input_ids because we're only matching strings at the end of the generated sequence
+        flipped_ids = torch.flip(input_ids, (1,))
+        string_matches = []
+        for stop_string in self.stop_strings:
+            target_len = len(stop_string)
+            max_valid_positions = self.max_valid_positions[stop_string]
+            max_valid_end_lens = self.max_valid_end_lens[stop_string]
+            # The embedding vec contains the valid positions, end_lengths and total lengths for each token
+            embedding_vec = self.embedding_vecs[stop_string]
+            embedded = F.embedding(flipped_ids, embedding_vec)
+
+            # Starts contains the number of characters from the string, counting from the end, that the token contains
+            # It can have multiple values if the same token can overlap different slices of the end of the string
+            starts = embedded[:, :1, max_valid_positions : max_valid_positions + max_valid_end_lens]
+            # Lengths is the total length of each token. Unlike starts, it always has a single value
+            lengths = embedded[:, 1:, -1:]
+            lengths = lengths.expand((-1, -1, starts.shape[-1]))
+            lengths_with_starts = torch.cat([starts, lengths], dim=1)
+            # We concatenate each possible starting length with the lengths of the remaining tokens in input_ids
+            # Then we cumsum() to get the total length of the string after each token
+            cumsum = lengths_with_starts.cumsum(dim=1)
+
+            # Valid positions are the positions in the string that the token can validly appear after
+            valid_positions = embedded[:, 1:, :max_valid_positions]
+            # Tokens can match the start of the string if they have any valid value in the starts vector
+            initial_match = torch.any(starts > 0, dim=-1, keepdim=True)
+            # Tokens can continue the string if the cumsum() so far is one of the valid positions for that token
+            # Note that we're actually tracking one cumsum() for the list for each possible start overhang length
+            later_match = torch.isin(cumsum[:, :-1], valid_positions)
+            # The match vector is a boolean vector that indicates which positions have valid tokens
+            match = torch.cat([initial_match, later_match], dim=1)
+
+            # Once a single position does not match, all positions following that position are masked
+            mask = (~match).cumsum(dim=1, dtype=torch.int32)
+            mask = mask == 0
+
+            # The string is matched if we reached a cumsum equal to or greater than the length of the string
+            # before hitting the mask
+            string_matches.append(torch.max(cumsum * mask, dim=1).values.squeeze(1) >= target_len)
+
+        # Now we concatenate the match booleans across all strings and check if any are True
+        string_matches = torch.cat(string_matches, dim=0)
+        return torch.any(string_matches).item()
```

Comment:
just being curious, so the generation stops for all batches when at least one has stop string?

Reply:
Yes, this is correct! I think this is the desired behaviour for all stopping conditions (cc @gante to confirm)